massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,317 @@
+pub(crate) struct DenunciationFactoryWorker {
+    cfg: FactoryConfig,
+    channels: FactoryChannels,
+    /// Factory manager command receiver
+    factory_receiver: Receiver<()>,
+    consensus_receiver: Receiver<DenunciationInterest>,
+    endorsement_pool_receiver: Receiver<DenunciationInterest>,
+    /// Internal cache for endorsement denunciation
+    /// store at most 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    endorsements_by_slot_index: HashMap<(Slot, u32), EndorsementDenunciationStatus>,
+    /// Internal cache for block header denunciation
+    block_header_by_slot: HashMap<Slot, BlockHeaderDenunciationStatus>,
+}
+
+impl DenunciationFactoryWorker {
+    /// Creates the `FactoryThread` structure to gather all data and references
+    /// needed by the factory worker thread.
+    pub(crate) fn spawn(
+        cfg: FactoryConfig,
+        channels: FactoryChannels,
+        factory_receiver: Receiver<()>,
+        consensus_receiver: Receiver<DenunciationInterest>,
+        endorsement_pool_receiver: Receiver<DenunciationInterest>,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("denunciation-factory".into())
+            .spawn(|| {
+                let mut factory = Self {
+                    cfg,
+                    channels,
+                    factory_receiver,
+                    consensus_receiver,
+                    endorsement_pool_receiver,
+                    endorsements_by_slot_index: Default::default(),
+                    block_header_by_slot: Default::default(),
+                };
+                factory.run();
+            })
+            .expect("failed to spawn thread : denunciation-factory")
+    }
+
+    /// Process new secured header (~ block header)
+    fn process_new_secured_header(
+        &mut self,
+        block_header_denunciation_interest: DenunciationInterest,
+    ) {
+        let de_i_orig = block_header_denunciation_interest.clone();
+        let de_i = match block_header_denunciation_interest {
+            DenunciationInterest::Endorsement(_) => {
+                return;
+            }
+            DenunciationInterest::BlockHeader(de_i) => de_i,
+        };
+
+        let now = MassaTime::now().expect("could not get current time");
+
+        // get closest slot according to the current absolute time
+        let slot_now = get_closest_slot_to_timestamp(
+            self.cfg.thread_count,
+            self.cfg.t0,
+            self.cfg.genesis_timestamp,
+            now,
+        );
+
+        let cycle_of_header = de_i.slot.get_cycle(self.cfg.periods_per_cycle);
+        let cycle_now = slot_now.get_cycle(self.cfg.periods_per_cycle);
+
+        // Do not fulfill the cache with block header too much in the future
+        // Note that cache is also purged each time a slot becomes final
+        if cycle_now - cycle_of_header > self.cfg.denunciation_items_max_cycle_delta {
```

Comment:
Problems:
* this does not do what the comment says
  * the comment says it rejects headers that are too much in the future
  * the code rejects headers that are too far in the past compare to now, which is not what we want
* artithmetic underflow if the header cycle is in the future => very important general rule: always think twice before using unprotected arithmetic operators
* lack of precision (granularity = cycles)

 Solutions:
* let's express all durations in periods instead of cycles
* first check if `de.slot.period < last_final_periods[de.slot.thread].saturating_sub(max_denunciation_periods) ` 
  * if true => ignore because the element is too old and cannot be denounced anymore anyway
* then check if `de.slot.period.saturating_sub(now.slot.period) > periods_per_cycle` 
  * if true => ignore because the element is too much in the future

Reply:
Fixed